src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.cs

```diff
@@ -0,0 +1,134 @@
+        /// diagnostic as well as for when running fix-all, then it should call 
+        /// <see cref="FixWithSyntaxEditorAsync"/> from its code action.  This will end up calling
+        /// <see cref="FixAllAsync"/>, with that single <paramref name="diagnostic"/> in the 
+        /// <see cref="ImmutableArray{T}"/> passed to that method.
+        /// </summary>
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeCodeActionsAsync(
+            Document document, Diagnostic diagnostic, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Subclasses should implement this to support fixing all given diagnostics efficiently.
+        /// </summary>
+        protected abstract Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics, SyntaxEditor editor, CancellationToken cancellationToken);
+
+        protected Task<Document> FixWithSyntaxEditorAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+            => SyntaxEditorBasedCodeFixProvider.FixAllWithEditorAsync(
+                document, editor => FixAllAsync(document, ImmutableArray.Create(diagnostic), editor, cancellationToken), cancellationToken);
+
+        #endregion
+
+        #region refactoring
+
+        /// <summary>
+        /// Subclasses should implement this to provide their feature as a refactoring.  This will be
+        /// called when the user has the code style enabled, but has set it to 'refactoring only'.
+        /// 
+        /// The implementation of this should check to see if the user's code is not in the form
+        /// specified by the option, and should offer to convert it to that form if possible.
+        /// </summary>
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeForwardRefactoringsAsync(
+            Document document, TextSpan span, TOptionKind option, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Subclasses should implement this to provide the refactoring that works in the opposite
+        /// direction of what the option preference is.  This is only called if the user has the
+        /// code style enabled, and has it set to 'info/warning/error'.  In this case it is the
+        /// *analyzer* responsible for making code compliant with the option.  
+        ///
+        /// The refactoring then exists to allow the user to update their code to go against that
+        /// option on an individual case by case basis.
+        ///
+        /// For example, if the user had set that they want expression-bodies for methods (at
+        /// warning level), then this would offer 'use block body' on a method that had an
+        /// expression body already.
+        /// </summary>
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeReverseRefactoringsAsync(
+            Document document, TextSpan span, TOptionKind option, CancellationToken cancellationToken);
```

Comment:
Note: i welcome different names than 'Forward' and 'Reverse' refactoring.  If anyone has any good ideas here, let me know!

Reply:
Perhaps:

```ComputeRefactoringsWhenAnalyzerNotPresent``` and
```ComputeOppositeRefactoringsWhenAnalyzerPresent```

?